dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/eventGraph/binary/MatIntersectionGraph.java

```diff
@@ -0,0 +1,87 @@
+package com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.binary;
+
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.EventGraph;
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.utils.MaterializedGraph;
+import com.dat3m.dartagnan.analysis.graphRefinement.util.GraphVisitor;
+import com.dat3m.dartagnan.verification.model.Edge;
+import com.dat3m.dartagnan.verification.model.ExecutionModel;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+// A materialized Union Graph. This seems to be more efficient than the virtualized UnionGraph
```

Comment:
First of all, the comment seems to be wrong (it mentions union instead of intersection). I will fix this.

The materialized version computes the explicit intersection and stores it, whereas the virtual intersection does not compute nor store anything upfront but instead computes the necessary data on demand. I originially used only the virtualized version, which is a lot more complicated and it seems also less performant in some cases. It is very memory efficient though. An ideal implementation would probably use both virtualized and materialized intersections, depending on the use case. However, currently I use the materialized version everywhere for ease of implementation.

In other words, we don't need the virtual version but there might be future use cases.